<img style="float: right; margin: 0px 0px 15px 15px;" src="https://www.topdoctors.es/files/Image/large/574d4503-26f8-433e-8e92-49355b793aaa.jpg" width="350px" height="180px" />

# <font color= #990066> Análisis Estadístico Multivariado </font>

- <Strong> `Claudia Celeste Castillejos Jáuregui` </Strong>
- <Strong> `5 abril 2022`</Strong>
- <Strong> `Practica #4`</Strong>
- <Strong> `claudia.castillejos@iteso.mx` </Strong> 
- <Strong> `Rocio Carrasco Navarro` </Strong> 

### <font color= #990066> Apéndice </font>

- Portada.
- Introducción.
- Descripción de la base de datos.
- Objetivo.
- Desarrollo.
- Resultados
- Conclusiones.

### <font color= #990066> Introducción </font>

En la práctica pasada utilice esta base de datos enfocada en seguros de vida. En este momento eliminaré esa columna y tomare como variable dependiente si fuma o no fuma y haré e análisis correspondiente con las demás variables independientes.

### <font color= #990066> Descripción de la base de datos </font>

Esta base de datos tiene 6 variables, 3 categóricas y 3 númericas.La variable dependiente es fumador, que se refiere a si fuma o no fuma este usuario 

Nombre de las variables:

* edad (age): edad del beneficiario principal

* sexo (sex): contratista de seguros género, femenino, masculino

* bmi (bmi): Índice de masa corporal, que proporciona una comprensión del cuerpo, pesos que son relativamente altos o bajos en relación con la altura,
índice objetivo de peso corporal (kg / m ^ 2) utilizando la relación entre altura y peso, idealmente 18,5 a 24,9

* niños (children): Número de niños cubiertos por el seguro de salud / Número de dependientes

* región (region): el área residencial del beneficiario en los EE. UU., noreste, sureste, suroeste, noroeste.

* fumador (smoker): si fuman (1) si no fuman (0)

Recuperado de:
https://www.kaggle.com/datasets/mirichoi0218/insurance


### <font color= #990066> Objectivo </font>

Realizar un ánalisis de resultados utilizando la regresión logistica y generar conclusiones correctas.

### <font color= #990066> Desarrollo </font>

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Cargamos la base de datos
data = pd.read_csv("insurance.csv", sep=',')
data

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

In [3]:
# Hacemos la variable númerica
data['smoker'].mask(data['smoker'] == 'no',0,inplace=True)
data['smoker'].mask(data['smoker'] == 'yes',1,inplace=True)
#Cambiamos el tipo de variable int64
data['smoker'] =data['smoker'].astype('int64')

In [4]:
# Borramos charges
data = data.drop(['charges'], axis=1)
data

age     sex     bmi  children  smoker     region
0      19  female  27.900         0       1  southwest
1      18    male  33.770         1       0  southeast
2      28    male  33.000         3       0  southeast
3      33    male  22.705         0       0  northwest
4      32    male  28.880         0       0  northwest
...   ...     ...     ...       ...     ...        ...
1333   50    male  30.970         3       0  northwest
1334   18  female  31.920         0       0  northeast
1335   18  female  36.850         0       0  southeast
1336   21  female  25.800         0       0  southwest
1337   61  female  29.070         0       1  northwest

[1338 rows x 6 columns]

In [5]:
# Checamos que se borrará completamente
data.columns.values

array(['age', 'sex', 'bmi', 'children', 'smoker', 'region'], dtype=object)


#### <font color= #990066> Ejercicio#1 </font>

In [6]:
#Creando la tabla de frecuencias 
tabla = pd.crosstab(index=data['smoker'], columns=data['region'],margins=True)
tabla

region  northeast  northwest  southeast  southwest   All
smoker                                                  
0             257        267        273        267  1064
1              67         58         91         58   274
All           324        325        364        325  1338

#### <font color= #990066> Resultados: </font> 

La tabla de frecuencias nos muestra que hay más personas que no fuman en comparación de las que si fuman por región. La región con más personas que fuman y que no fuman es southeast. La cantidad total de las personas que no fuman en esta muestra es 1064 y de las que si fuman 274.

In [7]:
# Aplicar logaritmos y medias de logaritmos
tab_log = np.log(tabla.iloc[0:-1,0:-1])
tab_log

region  northeast  northwest  southeast  southwest
smoker                                            
0        5.549076   5.587249   5.609472   5.587249
1        4.204693   4.060443   4.510860   4.060443

En está tabla se calculan los logaritmos y las medias logarítmicas de cada apartado

In [8]:
# Sacar las medias por columnas 
tab_log.loc['media_c'] = tab_log.mean(axis=0)
# Sacar las medias por filas 
tab_log['media_f'] = tab_log.mean(axis=1)

In [9]:
tab_log

region   northeast  northwest  southeast  southwest   media_f
smoker                                                       
0         5.549076   5.587249   5.609472   5.587249  5.583261
1         4.204693   4.060443   4.510860   4.060443  4.209110
media_c   4.876884   4.823846   5.060166   4.823846  4.896185

#### <font color= #990066> Resultados: </font> 

Esta es la tabla de los logaritmos y le agregamos las medias logarítmicas por columna y por filas. Están nos permite identificar si hay algún sesgo en los datos si están muy diferentes los resultados entre sí, en este caso no hay problema por las medias no cambian mucho.

In [10]:
# Calcular mu, mu1,mu2
# Seleccionar la media global de la tab_log
mu = tab_log.iloc[-1,-1]
mu

4.8961854179851665

In [11]:
# media por filas menos la media global 
mu1 = tab_log.iloc[:,-1]-mu
mu1

smoker
0          0.687076
1         -0.687076
media_c    0.000000
Name: media_f, dtype: float64

#### <font color= #990066> Resultados: </font> 

Para este ejemplo observamos que el mu1 fue negativo, esto significa ausencia por lo tanto la mayoria de la muestra no fuma.

In [12]:
# media por columnas menos la media global 
mu2 = tab_log.iloc[-1,:]-mu
mu2

region
northeast   -0.019301
northwest   -0.072340
southeast    0.163980
southwest   -0.072340
media_f      0.000000
Name: media_c, dtype: float64

In [13]:
# Tratar de recuperar los logaritmos 
tab_rec = np.ones(np.shape(tab_log))*mu
for k in np.arange(np.shape(tab_log)[0]):
    tab_rec[k,:] = tab_rec[k,:]+mu2
tab_rec

array([[4.87688435, 4.82384583, 5.06016565, 4.82384583, 4.89618542],
       [4.87688435, 4.82384583, 5.06016565, 4.82384583, 4.89618542],
       [4.87688435, 4.82384583, 5.06016565, 4.82384583, 4.89618542]])

In [14]:
for k in np.arange(np.shape(tab_log)[1]):
    tab_rec[:,k] = tab_rec[:,k]+mu1
tab_rec = tab_rec[0:-1,0:-1]
tab_rec

array([[5.56396023, 5.51092172, 5.74724153, 5.51092172],
       [4.18980847, 4.13676995, 4.37308977, 4.13676995]])

In [15]:
tab_mu12 = pd.DataFrame(np.transpose(np.transpose(tab_log-mu-mu2)-mu1))
tab_mu12.iloc[:,-1] = mu1
tab_mu12.iloc[-1,:] = mu2
tab_mu12

region   northeast  northwest  southeast  southwest   media_f
smoker                                                       
0        -0.014884   0.076327   -0.13777   0.076327  0.687076
1         0.014884  -0.076327    0.13777  -0.076327 -0.687076
media_c  -0.019301  -0.072340    0.16398  -0.072340  0.000000

In [16]:
tab_rec_mod = tab_rec+tab_mu12.iloc[0:-1,0:-1]
tab_rec_mod

region  northeast  northwest  southeast  southwest
smoker                                            
0        5.549076   5.587249   5.609472   5.587249
1        4.204693   4.060443   4.510860   4.060443

#### <font color= #990066> Resultados: </font> 

Nos da igual que la primera tabla lo que significa que se hizo bien la recuperación de los logaritmos 

In [17]:
# Analisis de las tablas de contigencia 
import scipy.stats as stats
# Analisis de 2 variables 
tabla1 = pd.crosstab(index=data['smoker'],columns=data['region'],margins=True)
# Ho: Las variables son independientes (p>0.05)
# Hi: Las variables no son independientes (p<0.05)
stats.chi2_contingency(tabla1)

(7.343477761407069,
 0.5000675325877666,
 8,
 array([[ 257.65022422,  258.44544096,  289.45889387,  258.44544096,
         1064.        ],
        [  66.34977578,   66.55455904,   74.54110613,   66.55455904,
          274.        ],
        [ 324.        ,  325.        ,  364.        ,  325.        ,
         1338.        ]]))

#### <font color= #990066> Resultados: </font> 

El chi cuadrada no es tan grande por lo tanto no se rechaza la hipotesis nula y p-value es mayor que 0.05  por lo tanto son independientes

#### <font color= #990066> Ejercicio#2 </font>

In [18]:
#Creando la tabla de frecuencias 
tabla = pd.crosstab(index=data['smoker'], columns=data['children'],margins=True)
tabla

children    0    1    2    3   4   5   All
smoker                                    
0         459  263  185  118  22  17  1064
1         115   61   55   39   3   1   274
All       574  324  240  157  25  18  1338

#### <font color= #990066> Resultados: </font> 

La tabla de frecuencias nos muestra que hay más personas que no fuman en comparación de las que si fuman por cantidad de hijos. La mayor cantidad de hijos que tienen las personas que no fuman es uno y para los que si fuman prefieren no tener hijos. La cantidad total de las personas que no fuman en esta muestra es 1064 y de las que si fuman 274 con respecto a cuantos hijos tienen.

In [19]:
# Aplicar logaritmos y medias de logaritmos
tab_log = np.log(tabla.iloc[0:-1,0:-1])
tab_log

children         0         1         2         3         4         5
smoker                                                              
0         6.129050  5.572154  5.220356  4.770685  3.091042  2.833213
1         4.744932  4.110874  4.007333  3.663562  1.098612  0.000000

En está tabla se calculan los logaritmos de cada apartado de la tabla.

In [20]:
# Sacar las medias por columnas 
tab_log.loc['media_c'] = tab_log.mean(axis=0)
# Sacar las medias por filas 
tab_log['media_f'] = tab_log.mean(axis=1)

In [21]:
tab_log

children         0         1         2         3         4         5   media_f
smoker                                                                        
0         6.129050  5.572154  5.220356  4.770685  3.091042  2.833213  4.602750
1         4.744932  4.110874  4.007333  3.663562  1.098612  0.000000  2.937552
media_c   5.436991  4.841514  4.613845  4.217123  2.094827  1.416607  3.770151

#### <font color= #990066> Resultados: </font> 

Esta es la tabla de los logaritmos y le agregamos las medias logarítmicas por columna y por filas. Están nos permite identificar si hay algún sesgo en los datos si están muy diferentes los resultados entre sí, en este caso no hay problema por las medias no cambian mucho.

In [22]:
# Calcular mu, mu1,mu2
# Seleccionar la media global de la tab_log
mu = tab_log.iloc[-1,-1]


In [23]:
# media por filas menos la media global 
mu1 = tab_log.iloc[:,-1]-mu
mu1

smoker
0          0.832599
1         -0.832599
media_c    0.000000
Name: media_f, dtype: float64

#### <font color= #990066> Resultados: </font> 

El negativo de el mu1 significa ausencia por lo tanto en esta muestra la mayoria no fuma.

In [24]:
# media por columnas menos la media global 
mu2 = tab_log.iloc[-1,:]-mu
mu2

children
0          1.666840
1          1.071363
2          0.843693
3          0.446972
4         -1.675324
5         -2.353544
media_f    0.000000
Name: media_c, dtype: float64

In [25]:
# Tratar de recuperar los logaritmos 
tab_rec = np.ones(np.shape(tab_log))*mu
for k in np.arange(np.shape(tab_log)[0]):
    tab_rec[k,:] = tab_rec[k,:]+mu2
tab_rec

array([[5.43699117, 4.84151395, 4.61384451, 4.21712314, 2.09482737,
        1.41660667, 3.77015113],
       [5.43699117, 4.84151395, 4.61384451, 4.21712314, 2.09482737,
        1.41660667, 3.77015113],
       [5.43699117, 4.84151395, 4.61384451, 4.21712314, 2.09482737,
        1.41660667, 3.77015113]])

In [26]:
for k in np.arange(np.shape(tab_log)[1]):
    tab_rec[:,k] = tab_rec[:,k]+mu1
tab_rec = tab_rec[0:-1,0:-1]
tab_rec

array([[6.26959012, 5.6741129 , 5.44644345, 5.04972208, 2.92742632,
        2.24920562],
       [4.60439222, 4.008915  , 3.78124556, 3.38452419, 1.26222842,
        0.58400772]])

In [27]:
tab_mu12 = pd.DataFrame(np.transpose(np.transpose(tab_log-mu-mu2)-mu1))
tab_mu12.iloc[:,-1] = mu1
tab_mu12.iloc[-1,:] = mu2
tab_mu12

children        0         1         2         3         4         5   media_f
smoker                                                                       
0        -0.14054 -0.101959 -0.226088 -0.279037  0.163616  0.584008  0.832599
1         0.14054  0.101959  0.226088  0.279037 -0.163616 -0.584008 -0.832599
media_c   1.66684  1.071363  0.843693  0.446972 -1.675324 -2.353544  0.000000

In [28]:
tab_rec_mod = tab_rec+tab_mu12.iloc[0:-1,0:-1]
tab_rec_mod

children         0         1         2         3         4         5
smoker                                                              
0         6.129050  5.572154  5.220356  4.770685  3.091042  2.833213
1         4.744932  4.110874  4.007333  3.663562  1.098612  0.000000

#### <font color= #990066> Resultados: </font> 

Da igual que la tabla de frecuencias por lo tanto si se hizo bien la recuperación de logaritmos.

In [29]:
# Analisis de las tablas de contigencia 
import scipy.stats as stats
# Analisis de 2 variables 
tabla1 = pd.crosstab(index=data['smoker'],columns=data['children'],margins=True)
# Ho: Las variables son independientes (p>0.05)
# Hi: Las variables no son independientes (p<0.05)
stats.chi2_contingency(tabla1)

(6.88771990494763,
 0.8649433793260529,
 12,
 array([[ 456.45440957,  257.65022422,  190.85201794,  124.8490284 ,
           19.88041854,   14.31390135, 1064.        ],
        [ 117.54559043,   66.34977578,   49.14798206,   32.1509716 ,
            5.11958146,    3.68609865,  274.        ],
        [ 574.        ,  324.        ,  240.        ,  157.        ,
           25.        ,   18.        , 1338.        ]]))

#### <font color= #990066> Resultados: </font> 

El chi cuadrada  es de 6.88771990494763 por lo tanto no se rechaza la hipotesis nula y p-value es mayor que 0.05 por lo tanto son independientes las variables analizadas previamente.

#### <font color= #990066> Ejercicio#3 </font>

In [30]:
#Creando la tabla de frecuencias 
tabla = pd.crosstab(index=data['smoker'], columns=data['sex'],margins=True)
tabla

sex     female  male   All
smoker                    
0          547   517  1064
1          115   159   274
All        662   676  1338

#### <font color= #990066> Resultados: </font> 

La tabla de frecuencias nos muestra que hay más personas que no fuman en comparación de las que si fuman entre hombres y mujeres. Las mujeres tienen un número más alto de personas que no fuman en comparación con los hombres.Los hombres tiene un número más alto de personas que si fuman en comparación con las mujeres. La cantidad total de las personas que no fuman en esta muestra es 1064 y de las que si fuman 274 con respecto a cuantos hijos tienen.

In [31]:
# Aplicar logaritmos y medias de logaritmos
tab_log = np.log(tabla.iloc[0:-1,0:-1])
tab_log

sex       female      male
smoker                    
0       6.304449  6.248043
1       4.744932  5.068904

En está tabla se calculan los logaritmos de cada apartado de la tabla.

In [32]:
# Sacar las medias por columnas 
tab_log.loc['media_c'] = tab_log.mean(axis=0)
# Sacar las medias por filas 
tab_log['media_f'] = tab_log.mean(axis=1)

In [33]:
tab_log

sex        female      male   media_f
smoker                               
0        6.304449  6.248043  6.276246
1        4.744932  5.068904  4.906918
media_c  5.524690  5.658474  5.591582

#### <font color= #990066> Resultados: </font> 

Esta es la tabla de los logaritmos y le agregamos las medias logarítmicas por columna y por filas. Están nos permite identificar si hay algún sesgo en los datos si están muy diferentes los resultados entre sí, en este caso no hay problema por las medias no cambian mucho.

In [34]:
# Calcular mu, mu1,mu2
# Seleccionar la media global de la tab_log
mu = tab_log.iloc[-1,-1]
mu

5.591582001878473

In [35]:
# media por filas menos la media global 
mu1 = tab_log.iloc[:,-1]-mu
mu1

smoker
0          0.684664
1         -0.684664
media_c    0.000000
Name: media_f, dtype: float64

#### <font color= #990066> Resultados: </font> 

El negativo del mu1 significa ausencia por lo tanto podemos decir que la mayoria no fuman

In [36]:
# media por columnas menos la media global 
mu2 = tab_log.iloc[-1,:]-mu
mu2

sex
female    -0.066892
male       0.066892
media_f    0.000000
Name: media_c, dtype: float64

In [37]:
# Tratar de recuperar los logaritmos 
tab_rec = np.ones(np.shape(tab_log))*mu
for k in np.arange(np.shape(tab_log)[0]):
    tab_rec[k,:] = tab_rec[k,:]+mu2
tab_rec

array([[5.52469047, 5.65847354, 5.591582  ],
       [5.52469047, 5.65847354, 5.591582  ],
       [5.52469047, 5.65847354, 5.591582  ]])

In [38]:
for k in np.arange(np.shape(tab_log)[1]):
    tab_rec[:,k] = tab_rec[:,k]+mu1
tab_rec = tab_rec[0:-1,0:-1]
tab_rec

array([[6.2093543 , 6.34313737],
       [4.84002663, 4.9738097 ]])

In [39]:
tab_mu12 = pd.DataFrame(np.transpose(np.transpose(tab_log-mu-mu2)-mu1))
tab_mu12.iloc[:,-1] = mu1
tab_mu12.iloc[-1,:] = mu2
tab_mu12

sex        female      male   media_f
smoker                               
0        0.095095 -0.095095  0.684664
1       -0.095095  0.095095 -0.684664
media_c -0.066892  0.066892  0.000000

In [40]:
tab_rec_mod = tab_rec+tab_mu12.iloc[0:-1,0:-1]
tab_rec_mod

sex       female      male
smoker                    
0       6.304449  6.248043
1       4.744932  5.068904

#### <font color= #990066> Resultados: </font> 

Nos da igual que la primera tabla lo que significa que se hizo bien la recuperación de los logaritmos 

In [41]:
# Analisis de las tablas de contigencia 
import scipy.stats as stats
# Analisis de 2 variables 
tabla1 = pd.crosstab(index=data['smoker'],columns=data['sex'],margins=True)
# Ho: Las variables son independientes (p>0.05)
# Hi: Las variables no son independientes (p<0.05)
stats.chi2_contingency(tabla1)

(7.765921028604452,
 0.10053907554919742,
 4,
 array([[ 526.43348281,  537.56651719, 1064.        ],
        [ 135.56651719,  138.43348281,  274.        ],
        [ 662.        ,  676.        , 1338.        ]]))

#### <font color= #990066> Resultados: </font> 

El chi 2 es de 7.765921028604452 ´por lo tanto se acepta la hipotesis nula y el p-value es de 0.10053907554919742 lo que significa que es mayor que 0.05 por lo tanto son independientes las variables

### <font color= #990066> Conclusiones: </font> 

Con referencia a los resultados que obtuvimos en esta práctica en los tres ejercicios logramos notar que las variables son independientes a la variable si fuma o no fuma el usuario. Considero que esta practica me ayudo a repasar lo conocimientos que obtuvimos en clase acerca de el tema Log-lineal 